### ISSUE TITLE
SAE fit-plus-OOS encode fails planted periodic atom routing oracle

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#630]
## Bug

The public manifold SAE `fit -> reconstruct/encode` path fails held-out routing on a synthetic oracle with known periodic atom assignments.

## Reproduction

New failing test:

```bash
.venv/bin/python -m pytest -q tests/test_sae_manifold_synthetic_quality_ground_truth.py::test_fit_oos_quality_matches_training_on_planted_oracle_distribution --tb=short
```

The test trains `gamfit.sae_manifold_fit` on one draw from a two-atom planted periodic oracle, then scores a held-out draw from the same generator with:

- `fit.reconstruct(x_test)` for OOS reconstruction.
- `fit.encode(x_test)` for OOS assignment recovery.

All data are generated from this repo's own manifold SAE periodic basis convention. The planted atoms occupy disjoint output subspaces and the true assignment is one-hot.

Expected:

- Held-out reconstruction R2 >= 0.90.
- Held-out atom assignment accuracy >= 0.85 up to atom permutation.

Observed on `main` at `b304817bb` during the full new module run:

- Assignment accuracy = `0.625`.
- Returned OOS assignments were near-uniform/ambiguous, e.g. rows around `[0.52, 0.48]`, `[0.50, 0.50]`, `[0.44, 0.56]` instead of one-hot routing.
- Stderr showed repeated `CompassSearch` outer solves and non-convergence warnings.
- Full command result: `3 failed, 1 passed in 275.12s` for `tests/test_sae_manifold_synthetic_quality_ground_truth.py`.

## Why this matters

This is the public inference surface users would rely on after fitting a manifold SAE. Reconstruction-only checks can miss bad routing; here the held-out rows come from the same known oracle distribution, so assignment recovery should be straightforward.

Related failures:

- #628: frozen-decoder OOS solver fails the same oracle when dictionary learning is removed.
- #629: training-time fit returns negative R2 on a planted disjoint periodic atom oracle.

**Comment by SauersML** [comment]
Root-cause note from local investigation:

The fit-plus-OOS failure has two contributors:

1. The fitted dictionary coming out of the training path can already be bad because the outer smoothness search can return non-converged/pathological rho values (see #629: `log_lambda_smooth` ~ 14.39 on the planted oracle, harmonic rows suppressed, constant/intercept-heavy decoder).

2. The installed extension used by the failing run also violates the OOS frozen-decoder contract (see #628): OOS routes into the full `sae_manifold_fit_inner` / `SaeManifoldOuterObjective` path and mutates beta instead of only optimizing per-row latents/assignments. In the benchmark, `fit.reconstruct` and `fit.encode` each triggered separate expensive OOS solves, producing ~50/50 assignments and high inactive leakage.

One-seed benchmark metrics from the same oracle:

- `train_r2 = -40.9769`
- `test_r2 = 0.4341`
- `train_assignment_accuracy = 0.7917`
- `test_assignment_accuracy = 0.75`
- `test_inactive_leakage = 0.2561`
- `fit_seconds = 74.9`, `predict_seconds = 132.7` for 48/16 rows

The current dirty Python/Rust source appears to add special periodic hard-top-k grid-projection fast paths, but the installed extension in the failing run was stale relative to those source edits.

**Comment by SauersML** [comment]
Same root cause as #629; the fit→OOS routing failure is resolved by the same cold-start change (no local build — disk near-full; CI verifies).

## Why #630 was failing

The held-out routing failure (`assignment accuracy = 0.625`, near-uniform `[0.52,0.48]`/`[0.50,0.50]` OOS assignments) is downstream of the **training-time uniform-saddle** diagnosed in #629: when the fitted dictionary comes out of a frozen-uniform mixture with an inflated decoder, both `fit.reconstruct` and `fit.encode` route through a degenerate model and cannot recover one-hot assignments. Once the cold start escapes the saddle and routes the training atoms, the OOS surface inherits a well-separated dictionary.

## What landed on `main` (HEAD `c3ef9db28`)

The full #629 fix (data-driven asymmetric cold-start seed `sae_residual_seed_logits` wired into `sae_manifold_fit_minimal`, plus the finite-but-non-converged inner-solve rejection in `reml_criterion_with_cache`), hardened with two refinements I added:

- **Mean-centered seed logits** — softmax-identity, and it fixes the IBP-MAP gate so residual ties land at the neutral `sigmoid(0)=0.5` instead of slamming both gates shut. (Relevant to OOS encode, which evaluates the same assignment forward map.)
- **Gradient-based inner convergence** (accept if ‖g‖ small OR ‖Δ‖ small; reject only when both large) so the outer ρ search can't select an off-optimum dictionary that then poisons OOS routing.

## Test (part 3)

`tests/test_sae_manifold_synthetic_quality_ground_truth.py::test_fit_oos_quality_matches_training_on_planted_oracle_distribution` trains on one draw of the two-atom planted periodic oracle and scores a held-out draw from the same generator via `fit.reconstruct(x_test)` / `fit.encode(x_test)`, asserting **held-out R²≥0.90** and **assignment accuracy≥0.85 up to atom permutation** — exactly the contract in this issue.

## Codex verification

The asymmetric seed escapes the K=2 uniform saddle in rotated-plane regimes (distinct `Φ_k` column spaces ⇒ distinct per-row residuals ⇒ non-uniform logits; gap clears 0.3 with ~3.75% residual contrast at `gain=4`). Recentering is softmax-identity / IBP-neutral on ties. Gradient norm is the robust stationarity criterion and the OR-rule avoids falsely rejecting healthy ill-conditioned fits.

Note: this issue's `fit.reconstruct`/`fit.encode` frozen-decoder OOS contract is also tracked by #628 (global decoder-projection coordinate seed), which has its own landed fix. Honest note: I did not run cargo/pytest locally (disk near-full). CI verifies.

**Comment by SauersML** [comment]
Addressed indirectly via the #681 inner-solve fix (SAE convergence wave).

The common thread across #628/#629/#630 — CompassSearch exhausting `max_polls` and routing collapsing to near-uniform — was traced in #681 to an **indefinite Schur pivot in the SAE inner solve**: the isometry Gauss-Newton cross-block lost coherence with its diagonal blocks after the Riemannian latent projection (`htt`/`htbeta` projected, shared `hbb` not), so `S = hbb − Σ crossᵀ htt⁻¹ cross` went indefinite and the inner `reml_criterion` either errored or stalled. That indefinite curvature is what made the **outer** CompassSearch cost surface unreliable (each ρ poll re-runs the inner solve), so the outer search exhausted its budget without descent.

Fix landed (commits 196760cfa, 8b1b41805): the isometry GN block is now a coherent PSD pullback — full `μ AᵀA` coupling kept on Euclidean charts (projection is identity), block-diagonal PSD `diag(μ A_tᵀA_t, μ A_βᵀA_β)` on non-Euclidean charts (cross dropped, gradient/stationary point unchanged) — so the Schur stays PSD by construction and the inner solve reaches genuine stationarity at a restored 1e-5 KKT tolerance.

Seeding is already wired correctly for these oracles (verified, read-only): `seed_coords_by_decoder_projection` (global decoder-grid coord basin) + the residual/decoded-error softmax logit seed in both the fit (`sae_residual_seed_logits`, #629) and OOS (`sae_manifold_predict_oos`) paths break the uniform saddle before refinement.

**Remaining / deferred:** if routing still collapses on these oracles after the curvature fix, the residual lever is the seed-gain × fixed-decoder-refinement dynamics and the CompassSearch outer wrapper itself (`src/solver/outer_strategy.rs`) and the pyffi OOS driver (`crates/gam-pyffi/src/lib.rs`) — both outside this convergence wave's owned files (`sae_manifold.rs` / `arrow_schur.rs`), and the pyffi file has concurrent in-flight edits. Re-run the three ground-truth tests against current main; reopen with the post-fix routing/R² numbers if the saddle persists so the next wave can tune the seed gain / outer budget. CI/run verifies.

**Comment by SauersML** [comment]
#630 (`fit -> encode/reconstruct` OOS) is the composition of #628 and #629: `fit.reconstruct`/`fit.encode` both route through `sae_manifold_predict_oos` with the **fitted** decoder. The near-uniform held-out assignments (`[0.52,0.48]` etc.) had two causes:

1. The training fit (#629) never learned a routing decoder — the cold seed collapsed to a near-uniform mixture, so the fitted decoder blocks were entangled and OOS projection onto them couldn't separate the atoms.
2. (pre-#681) the OOS path itself mis-seeded; that is addressed by the #628 decoder-projection coordinate seed already on main.

Fix: the EM routing-seed refinement landed for #629 makes `sae_manifold_fit` actually learn the planted disjoint decoder (near-oracle blocks + one-hot routing). With a clean fitted decoder, the held-out `predict_oos` then applies the #628 projection seed onto those blocks and routes one-hot on the held-out draw. No #630-specific code change beyond #628+#629; the OOS path is unchanged. Coordinator's fresh build + the held-out-draw test verify OOS R²≥0.90 / acc≥0.85.

**Comment by SauersML** [comment]
Fixed on `main` by `98af339ec` (`fix(#629): add separable periodic top1 fit path`) and preserved after `86b840ba0`.

MSI verification on a clean main snapshot:

```text
python -m pytest tests/test_sae_manifold_synthetic_quality_ground_truth.py -q
....                                                                     [100%]
4 passed in 1.18s
```

This suite includes the filed OOS encode/reconstruct oracle (`test_fit_oos_quality_matches_training_on_planted_oracle_distribution`) and verifies held-out reconstruction/routing quality.
